### PR TITLE
Add partner autocomplete to communication record form [PD-3810]

### PR DESCRIPTION
**Pull request by tdphillips**
## What's this PR do?
Adds a Partner autocomplete to the communication record form if you reach it from the common action button.

We also remove waffle switches as the switched feature has been released.

## Screenshots (if appropriate)
2017-09-27-152452_801x257_scrot

## Where should the reviewer start?
Most of the code is in `mypartners/forms.py`. A bit more is in `mypartners/views.py`.

## How should this be manually tested?
Go to PRM. Click the button. Open a partner. Click the button. Go to NUO and Partner Assessment. Click the button on each.

### Front-End

- [ ] Works on IE 11 and Edge
- [ ] Works on Mobile
- [ ] Unit tests added or ticket exists to add tests

### Separation of Concerns

- [ ] Back end "business logic" is implemented in plain python, Django model classes, and Django forms. (not Django views or other frameworks)
- [ ] Concerns of UI, api, and model are in separate code units.

### Tests

- [ ] Base classes used for python tests don't do too much unnecessary work.

### Migrations

- [ ] Migrations can be reversed.
- [ ] Current production code would continue to work after this migration is applied.
- [ ] Data migrations' forward/reverse functions should be decorated with `@only_on_default`.

### CODE REVIEW

File: gulp/src/partnerassessment/components/PartnerAssessmentApp.jsx [hunk 1]
```diff
@@ -52,6 +42,16 @@ class PartnerAssessmentApp extends Component {
   }
   render() {
     const {partnerName, partnerId, primaryContact} = this.props;
+    const commonAction = (
```
Comment (tdphillips) [inline gulp/src/partnerassessment/components/PartnerAssessmentApp.jsx]: We do this in `render()` in `PATertiaryNavigation.jsx` so I assume that's okay here as well.

File: mypartners/forms.py [hunk 2]
```diff
@@ -364,33 +366,81 @@ def save(self, request, commit=True):
         return instance
 
 
+class HorizontalRadioRenderer(forms.RadioSelect.renderer):
```
Comment (tdphillips) [inline mypartners/forms.py]: This has moved. Nothing was changed.

File: mypartners/forms.py [hunk 3]
```diff
@@ -364,33 +366,81 @@ def save(self, request, commit=True):
...
+                id='%s_%s' % (self.attrs['id'], choice[0]),
+                name=self.name,
+                value=choice[0],
+                title=self.radio_tooltips.get(choice[0], '')
+            ))
+        return mark_safe(u'\n'.join([u'%s\n' % w for w in radios]))
+
+
+@register('partner')
+class PartnerLookup(LookupChannel):
```
Comment (tdphillips) [inline mypartners/forms.py]: Basically straight from the django-ajax-selects docs.

File: mypartners/forms.py [hunk 4]
```diff
@@ -364,33 +366,81 @@ def save(self, request, commit=True):
...
+        return item.name
+
+
 class ContactRecordForm(NormalizedModelForm):
     date_time = SplitDateTimeDropDownField(label='Date & Time')
     length = TimeDropDownField()
     attachment = MultipleFileField(required=False,
                                    help_text="Max file size %sMB" %
                                              MAX_ATTACHMENT_MB)
+    score = forms.ChoiceField(
```
Comment (tdphillips) [inline mypartners/forms.py]: This looks like an addition but really it's replacing `ContactRecordForm` with `ContactRecordFormWithAssessment`

File: templates/mypartners/edit_record.html [hunk 5]
```diff
@@ -47,7 +53,6 @@
                 {% include "includes/form-error-highlight.html" %}
                 {% csrf_token %}
                 <input type="hidden" name="company" value="{{ company.id }}">
-                <input type="hidden" name="partner" value="{{ partner.id }}">
```
Comment (tdphillips) [inline templates/mypartners/edit_record.html]: We're including `partner` in the form now. We don't need to fake it here.

File: templates/mypartners/edit_record.html [hunk 6]
```diff
@@ -85,6 +90,8 @@
 {% endblock %}
 
 {% block extra-js %}
+    {{ form.media }}
```
Comment (tdphillips) [inline templates/mypartners/edit_record.html]: django-ajax-selects requires some JS.

**Comment by tdphillips** [comment]
This needs some CSS work.

**Comment by tdphillips** [comment]
I need to hide some fields and do ajax shenanigans.

**Comment by tdphillips** [comment]
There's a bunch of jQuery. Read it and weep.

Proof of concept. If we don't want to do this, no problem.

File: static/partner_edit_record.192-28.js [hunk 7]
```diff
@@ -1,10 +1,50 @@
...
+
 $(function() {
+    $("#prm-application-body").on("click", "[id^='kill_'][id$='id_partner']", function() {
+        $("input#id_partner_text").show();
+        hide_form();
+    });
+
+    $("#prm-application-body").on("click", ".ui-autocomplete li", function() {
+        $("input#id_partner_text").hide();
+        populate_contacts($('#id_partner').val());
```
Comment (tdphillips) [inline static/partner_edit_record.192-28.js]: This should probably be memoized somewhere so we don't retrieve the same list of contacts multiple times. I'll get to that soon.

**Comment by koonsies** [comment]
Great job with the look and feel of the controls! During my review, I found that the following workflow throws an error

1) Select a partner from the partner list
2) Click the common action button
3) Click the "trash" to remove the partner as the selected partner
4) Select another partner
5) Fill out the form as usual
6) Save and see the failed validation on contact.

**Comment by tdphillips** [comment]
Whoops. Thanks. I should check more than just the happy path.

**Comment by tdphillips** [comment]
Sorry, that change was stashed. I pushed it. There are still some tests to work on, however.

**Comment by koonsies** [comment]
This is weird behavior and I don't know what to make of it. If you create a new CommRec via the common action button, then immediately click "Edit" on the CommRec summary screen shown, you can change the partner on the form still. If you click back to Partners, select the same partner and go to that communication record summary, edit does not allow changing of the partner.

Not sure why this is, but I assume we don't want those two to be inconsistent.

**Comment by koonsies** [comment]
Oh, it's because "common" remains in the url after saving the form.

File: myreports/tests/test_views.py [hunk 8]
```diff
@@ -548,6 +553,7 @@ def test_select_data_type_api(self):
             u'selected_report_category': u'communication-records',
             u'selected_report_intention': u'prm',
         }
+        import ipdb; ipdb.set_trace()
```
Comment (koonsies) [inline myreports/tests/test_views.py]: ipdb statement

Comment (tdphillips) [reply myreports/tests/test_views.py]: It's been fixed.

**Pull request merged**